[BUG] BatchWrite BatchGet StackOverflowError

**Describe the bug**
<!-- A clear and concise description of what the bug is. -->
```
Caused by: java.lang.StackOverflowError
        at java.util.TreeMap.compare(TreeMap.java:1294)
        at java.util.TreeMap.getFloorEntry(TreeMap.java:429)
        at java.util.TreeMap.floorEntry(TreeMap.java:722)
        at com.pingcap.com.google.common.collect.TreeRangeMap.getEntry(TreeRangeMap.java:113)
        at com.pingcap.com.google.common.collect.TreeRangeMap.get(TreeRangeMap.java:105)
        at com.pingcap.tikv.region.RegionManager$RegionCache.getRegionByKey(RegionManager.java:203)
        at com.pingcap.tikv.region.RegionManager.getRegionByKey(RegionManager.java:72)
        at com.pingcap.tikv.region.RegionManager.getRegionByKey(RegionManager.java:68)
        at java.util.stream.Collectors.lambda$groupingBy$45(Collectors.java:907)
        at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at com.pingcap.tikv.KVClient.groupKeysByRegion(KVClient.java:245)
        at com.pingcap.tikv.KVClient.doSendBatchGetWithRefetchRegion(KVClient.java:195)
        at com.pingcap.tikv.KVClient.doSendBatchGetInBatchesWithRetry(KVClient.java:189)
        at com.pingcap.tikv.KVClient.doSendBatchGetWithRefetchRegion(KVClient.java:205)
        at com.pingcap.tikv.KVClient.doSendBatchGetInBatchesWithRetry(KVClient.java:189)
        at com.pingcap.tikv.KVClient.doSendBatchGetWithRefetchRegion(KVClient.java:205)
        at com.pingcap.tikv.KVClient.doSendBatchGetInBatchesWithRetry(KVClient.java:189)
        at com.pingcap.tikv.KVClient.doSendBatchGetWithRefetchRegion(KVClient.java:205)
        at com.pingcap.tikv.KVClient.doSendBatchGetInBatchesWithRetry(KVClient.java:189)
        at com.pingcap.tikv.KVClient.doSendBatchGetWithRefetchRegion(KVClient.java:205)
        at com.pingcap.tikv.KVClient.doSendBatchGetInBatchesWithRetry(KVClient.java:189)
        at com.pingcap.tikv.KVClient.doSendBatchGetWithRefetchRegion(KVClient.java:205)
        at com.pingcap.tikv.KVClient.doSendBatchGetInBatchesWithRetry(KVClient.java:189)
        at com.pingcap.tikv.KVClient.doSendBatchGetWithRefetchRegion(KVClient.java:205)
```

**What did you do**
<!-- 
If possible, please provide a code receipt to produce this issue.
e.g., 
1. run `spark-shell --jars xxx.jar`
2. in spark-shell, run 
```
val df = spark.sql(“select sum(b) from t where a = 1”)
df.show(20)
```
3. update some data through TiDB
4. in spark-shell, run
```
val df = spark.sql(“select sum(b) from t where a = 1”)
df.show(20)
```
 -->

**What do you expect**
<!-- A clear and concise description of what you expected to happen. -->

**What happens instead**
<!-- If an error occurs, please provide complete error stack. -->

<!-- 
**Screenshots**
If applicable, add screenshots to help explain your problem.
 -->

**Spark and TiSpark version info**
<!-- What version of Spark and TiSpark are you using? (Provide Spark version and run `spark.sql(“select ti_version()”).show(false)` in spark-shell) -->
tispark-2.3.7

<!--
**Additional context**
Add any other context about the problem here.
You may also provide TiDB version here if it is related to the issue.
-->

--- a/tikv-client/src/main/java/com/pingcap/tikv/KVClient.java
+++ b/tikv-client/src/main/java/com/pingcap/tikv/KVClient.java
@@ -18,7 +18,7 @@
 package com.pingcap.tikv;
 
 import static com.pingcap.tikv.util.ClientUtils.getBatches;
-import static com.pingcap.tikv.util.ClientUtils.getKvPairs;
+import static com.pingcap.tikv.util.ClientUtils.getTasksWithOutput;
 
 import com.google.protobuf.ByteString;
 import com.pingcap.tikv.exception.GrpcException;
@@ -30,9 +30,12 @@ import com.pingcap.tikv.util.BackOffFunction;
 import com.pingcap.tikv.util.BackOffer;
 import com.pingcap.tikv.util.Batch;
 import com.pingcap.tikv.util.ConcreteBackOffer;
+import com.pingcap.tikv.util.Pair;
 import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Objects;
+import java.util.Queue;
 import java.util.concurrent.ExecutorCompletionService;
 import java.util.concurrent.ExecutorService;
 import org.slf4j.Logger;
@@ -94,55 +97,56 @@ public class KVClient implements AutoCloseable {
   }
 
   private List<KvPair> doSendBatchGet(BackOffer backOffer, List<ByteString> keys, long version) {
-    ExecutorCompletionService<List<KvPair>> completionService =
+    ExecutorCompletionService<Pair<List<Batch>, List<KvPair>>> completionService =
         new ExecutorCompletionService<>(batchGetThreadPool);
 
     List<Batch> batches =
         getBatches(backOffer, keys, BATCH_GET_SIZE, MAX_BATCH_LIMIT, this.clientBuilder);
 
-    for (Batch batch : batches) {
-      completionService.submit(
-          () -> doSendBatchGetInBatchesWithRetry(batch.getBackOffer(), batch, version));
+    // prevent stack overflow
+    Queue<List<Batch>> taskQueue = new LinkedList<>();
+    List<KvPair> result = new ArrayList<>();
+    taskQueue.offer(batches);
+
+    while (!taskQueue.isEmpty()) {
+      List<Batch> task = taskQueue.poll();
+      for (Batch batch : task) {
+        completionService.submit(
+            () -> doSendBatchGetInBatchesWithRetry(batch.getBackOffer(), batch, version));
+      }
+      result.addAll(
+          getTasksWithOutput(completionService, taskQueue, task, BackOffer.RAWKV_MAX_BACKOFF));
     }
 
-    return getKvPairs(completionService, batches, BackOffer.BATCH_GET_MAX_BACKOFF);
+    return result;
   }
 
-  private List<KvPair> doSendBatchGetInBatchesWithRetry(
+  private Pair<List<Batch>, List<KvPair>> doSendBatchGetInBatchesWithRetry(
       BackOffer backOffer, Batch batch, long version) {
     TiRegion oldRegion = batch.getRegion();
     TiRegion currentRegion =
-        clientBuilder.getRegionManager().getRegionByKey(oldRegion.getStartKey());
-
+        clientBuilder.getRegionManager().getRegionByKey(batch.getRegion().getStartKey());
     if (oldRegion.equals(currentRegion)) {
       RegionStoreClient client = clientBuilder.build(batch.getRegion());
       try {
-        return client.batchGet(backOffer, batch.getKeys(), version);
+        List<KvPair> partialResult = client.batchGet(backOffer, batch.getKeys(), version);
+        return Pair.create(new ArrayList<>(), partialResult);
       } catch (final TiKVException e) {
         backOffer.doBackOff(BackOffFunction.BackOffFuncType.BoRegionMiss, e);
         clientBuilder.getRegionManager().invalidateRegion(batch.getRegion());
-        logger.warn("ReSplitting ranges for BatchGetRequest", e);
-
-        // retry
-        return doSendBatchGetWithRefetchRegion(backOffer, batch, version);
+        logger.debug("ReSplitting ranges for BatchGetRequest", e);
+        return doRetryBatchGet(backOffer, batch);
       }
     } else {
-      return doSendBatchGetWithRefetchRegion(backOffer, batch, version);
+      return doRetryBatchGet(backOffer, batch);
     }
   }
 
-  private List<KvPair> doSendBatchGetWithRefetchRegion(
-      BackOffer backOffer, Batch batch, long version) {
-    List<Batch> retryBatches =
-        getBatches(backOffer, batch.getKeys(), BATCH_GET_SIZE, MAX_BATCH_LIMIT, this.clientBuilder);
+  private Pair<List<Batch>, List<KvPair>> doRetryBatchGet(BackOffer backOffer, Batch batch) {
+    return Pair.create(doSendBatchGetWithRefetchRegion(backOffer, batch), new ArrayList<>());
+  }
 
-    ArrayList<KvPair> results = new ArrayList<>();
-    for (Batch retryBatch : retryBatches) {
-      // recursive calls
-      List<KvPair> batchResult =
-          doSendBatchGetInBatchesWithRetry(retryBatch.getBackOffer(), retryBatch, version);
-      results.addAll(batchResult);
-    }
-    return results;
+  private List<Batch> doSendBatchGetWithRefetchRegion(BackOffer backOffer, Batch batch) {
+    return getBatches(backOffer, batch.getKeys(), BATCH_GET_SIZE, MAX_BATCH_LIMIT, clientBuilder);
   }
 }

